### DISCUSSION TITLE
Consuming the STL module with CMake

### DISCUSSION THREAD

**Discussion by Zingam** [microsoft/STL#3996] [Q&A] [answered]
https://learn.microsoft.com/en-us/cpp/build/reference/c-cpp-prop-page?view=msvc-170#build-iso-c23-standard-library-modules
https://learn.microsoft.com/en-us/cpp/cpp/tutorial-import-stl-named-module?view=msvc-170

It seems that consuming the STL module form CMake isn't something out-of-the-box? 
What would I need to do to setup a CMake project with CMakePresets to be able to consume the STL module and this should happen crossplatform with Clang/GCC?

**Comment by StephanTLavavej** [comment]
We've heard that Kitware is working on making this work out-of-the-box with CMake, but we can't share any ETA at this time.

**Comment by fsb4000** [comment]
@Zingam

Have you seen this: https://github.com/microsoft/STL/issues/3195#issuecomment-1770283053 ?

**Comment by davidhunter22** [comment]
using `import std;` seems to work fine with Visual Studio solution/projects generated from CMake.
You do need to set the `CXX_SCAN_FOR_MODULES` property to true in your CMake. this controls the `ScanSourceForModuleDependencies` property in the generated `.vcxproj` files.
As of right now the `.vcxproj` generation is a bit of a mess, see https://gitlab.kitware.com/cmake/cmake/-/issues/25806, but it does work.
Note you can't control the related `BuildStlModules` MSBuild property directly from CMake, although quite why that property exists is a bit of a mystery to me!

**Comment by StephanTLavavej** [comment, marked as answer]
https://www.kitware.com/import-cmake-the-experiment-is-over/

> At last the experiment is over, and CMake 3.28 has official support for C++ 20 named modules enabled without having to set the CMAKE_EXPERIMENTAL_CXX_MODULE_CMAKE_API variable. This has been a long adventure involving the coordination of the SG15 tooling group, and developers of three major compilers: Visual Studio, LLVM/Clang, and GCC. To read about the entire process and the technical details of how CMake supports named modules see this blog: https://www.kitware.com/import-cmake-c20-modules/.

> C++ 20 named modules are now supported in CMake 3.28 and newer when using  Ninja and Visual Studio Generators (VS 2022 and newer in combination with the MSVC 14.34 toolset, provided with VS 17.4, and newer), LLVM/Clang 16.0 and newer, and GCC 14 (after the 2023-09-20 daily bump) and newer.
